core/modules/replica/WorkerSqlRequest.cc

```diff
@@ -0,0 +1,217 @@
+                             query,
+                             user,
+                             password,
+                             maxRows));
+}
+
+
+WorkerSqlRequest::WorkerSqlRequest(ServiceProvider::Ptr const& serviceProvider,
+                                   string const& worker,
+                                   string const& id,
+                                   int priority,
+                                   std::string const& query,
+                                   std::string const& user,
+                                   std::string const& password,
+                                   size_t maxRows)
+    :   WorkerRequest(serviceProvider,
+                      worker,
+                      "SQL",
+                      id,
+                      priority),
+        _query(query),
+        _user(user),
+        _password(password),
+        _maxRows(maxRows) {
+}
+
+
+void WorkerSqlRequest::setInfo(ProtocolResponseSql& response) const {
+
+    LOGS(_log, LOG_LVL_DEBUG, context(__func__));
+
+    util::Lock lock(_mtx, context(__func__));
+
+    // Update the performance of the target request before returning it
+    response.set_allocated_target_performance(performance().info());
+
+    // Carry over the result of the query only after the request
+    // has finished (or failed).
+    switch (status()) {
+        case STATUS_SUCCEEDED:
+        case STATUS_FAILED:
+            response.set_error(            _response.error());
+            response.set_has_result(       _response.has_result());
+            *(response.mutable_fields()) = _response.fields();
+            *(response.mutable_rows())   = _response.rows();
+            break;
+        default:
+            break;
+    }
+}
+
+
+bool WorkerSqlRequest::execute() {
+
+    LOGS(_log, LOG_LVL_DEBUG, context(__func__));
+
+    util::Lock lock(_mtx, context(__func__));
+
+    switch (status()) {
+
+        case STATUS_IN_PROGRESS:
+            break;
+
+        case STATUS_IS_CANCELLING:
+
+            // Abort the operation right away
+
+            setStatus(lock, STATUS_CANCELLED);
+            throw WorkerRequestCancelled();
+
+        default:
+            throw logic_error(
+                    "WorkerSqlRequest::" + context(__func__) + "  not allowed while in state: " +
+                    WorkerRequest::status2string(status()));
+    }
+
+    database::mysql::Connection::Ptr conn;
+    try {
+        auto const workerInfo = serviceProvider()->config()->workerInfo(worker());
+        conn = database::mysql::Connection::open(
+            database::mysql::ConnectionParams(
+                workerInfo.dbHost,
+                workerInfo.dbPort,
+                user(),
+                password(),
+                ""));
+
+        auto self = shared_from_base<WorkerSqlRequest>();
+        conn->execute([self](decltype(conn) const& conn_) {
+            conn_->begin();
+            conn_->execute(self->query());
```

Comment:
It could make sense building query outside transaction if it has chance to throw.

Reply:
This is kust a trivial non-throwing accessor method. The query is supposed to be prepared.

Reply:
How do I know it here? What if `query()` implementation changes?